libnavigation-core/src/test/java/com/mapbox/navigation/core/routerefresh/RouteRefreshControllerTest.kt

```diff
@@ -1,307 +1,690 @@
-        every { routeOptions.enableRefresh() } returns true
+    fun `should refresh route with any annotation`() = coroutineRule.runBlockingTest {
+        val routeWithoutAnnotations = createTestTwoLegRoute(
+            firstLegAnnotations = null,
+            secondLegAnnotations = null
+        )
+        val refreshedRoute = createNavigationRoute()
+        val directionsSession = mockk<DirectionsSession>().apply {
+            onRefresh { _, _, _, callback -> callback.onRefreshReady(refreshedRoute) }
+        }
+        val routeRefreshController = createRouteRefreshController(
+            directionsSession = directionsSession
+        )
 
-        routeRefreshController.restart(validRoute) {}
-        coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
-        routeRefreshController.stop()
+        val refreshedRouteDeferred =
+            async { routeRefreshController.refresh(listOf(routeWithoutAnnotations)) }
+        advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
 
+        assertEquals(listOf(refreshedRoute), refreshedRouteDeferred.getCompletedTest())
         verify(exactly = 1) { directionsSession.requestRouteRefresh(any(), any(), any()) }
     }
 
-    @Test
-    fun `should refresh route with any annotation`() =
-        coroutineRule.runBlockingTest {
-            every { routeOptions.enableRefresh() } returns true
-
-            routeRefreshController.restart(validRoute) {}
-            coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
-            routeRefreshController.stop()
-
-            verify(exactly = 1) { directionsSession.requestRouteRefresh(any(), any(), any()) }
-        }
-
     @Test
     fun `should log warning when route is not supported`() = coroutineRule.runBlockingTest {
-        every { routeOptions.enableRefresh() } returns true
-        every { validRoute.directionsResponse.uuid() } returns null
+        val primaryRoute = createTestTwoLegRoute(requestUuid = null)
+        val directionsSession = mockk<DirectionsSession>()
+        val routeRefreshController = createRouteRefreshController()
 
-        routeRefreshController.restart(validRoute) {}
-        coroutineRule.testDispatcher.advanceTimeBy(TimeUnit.MINUTES.toMillis(6))
-        routeRefreshController.stop()
+        val refreshedDeferred = async { routeRefreshController.refresh(listOf(primaryRoute)) }
+        advanceTimeBy(TimeUnit.HOURS.toMillis(6))
 
+        assertTrue(refreshedDeferred.isActive)
         verify(exactly = 0) { directionsSession.requestRouteRefresh(any(), any(), any()) }
         verify(exactly = 1) {
-            logger.logW(any(), any())
+            logger.logI(any(), any())
         }
+        refreshedDeferred.cancel()
     }
 
     @Test
-    fun `cancel request when stopped (nothing started)`() {
-        routeRefreshController.stop()
+    fun `refreshing of empty routes`() = coroutineRule.runBlockingTest {
+        val primaryRoute = createTestTwoLegRoute(requestUuid = null)
```

Comment:
Does this test actually test empty routes? Or does it succeed because the UUID is `null`?

Reply:
it just checks UUID.
Can you please tell me more about this scenario. The test you see is just ported from callback to suspend function previously existing test.

Reply:
oh, I've just realised what is wrong 😱 
I wanted to test empty rotes!

Reply:
fixed.